src/ir_Teco.cpp

```diff
@@ -0,0 +1,264 @@
+// Copyright 2019 Fabien Valthier
+/*
+Node MCU/ESP8266 Sketch to emulate Teco
+*/
+
+#include "ir_Teco.h"
+#include <algorithm>
+#include "IRremoteESP8266.h"
+#include "IRutils.h"
+#ifndef ARDUINO
+#include <string>
+#endif
+
+// Constants
+// using SPACE modulation.
+const uint16_t kTecoHdrMark = 9000;
+const uint16_t kTecoHdrSpace = 4440;
+const uint16_t kTecoBitMark = 620;
+const uint16_t kTecoOneSpace = 1650;
+const uint16_t kTecoZeroSpace = 580;
+
+#if SEND_TECO
+// Send a Teco A/C message.
+//
+//
+// Args:
+//   data:   Contents of the message to be sent.
+//   nbits:  Nr. of bits of data to be sent. Typically kTecoBits.
+//   repeat: Nr. of additional times the message is to be sent.
+
+void IRsend::sendTeco(uint64_t data, uint16_t nbits, uint16_t repeat) {
+  // if (nbits % 35 != 0) return;  // Not enough bits to send a proper message.
+  sendGeneric(kTecoHdrMark, kTecoHdrSpace, kTecoBitMark, kTecoOneSpace,
+              kTecoBitMark, kTecoZeroSpace, kTecoBitMark, 0,
+              data, nbits, 38, false, repeat, kDutyDefault);
+}
+#endif  // SEND_TECO
+
+IRTecoAC::IRTecoAC(uint16_t pin) : _irsend(pin) { stateReset(); }
+
+void IRTecoAC::begin() { _irsend.begin(); }
+
+#if SEND_TECO
+void IRTecoAC::send(const uint16_t repeat) {
+  _irsend.sendTeco(remote_state, kTecoBits, repeat);
+}
+#endif  // SEND_TECO
+
+void IRTecoAC::stateReset() {
+  // automatically sets: Mode:auto, Off, fan:auto, temp:16
+  remote_state = kTecoReset;
+  this->setSwing(true);
+}
+
+uint64_t IRTecoAC::getRaw() { return remote_state; }
+
+void IRTecoAC::setRaw(const uint64_t new_code) { remote_state = new_code; }
+
+void IRTecoAC::on() {
+  remote_state |= kTecoPower;
+}
+
+void IRTecoAC::off() {
+  remote_state &= ~kTecoPower;
+}
+
+void IRTecoAC::setPower(bool state) {
+  if (state)
+    on();
+  else
+    off();
+}
+
+bool IRTecoAC::getPower() {
+  return (remote_state & kTecoPower) == kTecoPower; }
+
+void IRTecoAC::setTemp(uint8_t temp) {
+  uint8_t newtemp = temp;
+  newtemp = std::min(newtemp, kTecoMaxTemp);
+  newtemp = std::max(newtemp, kTecoMinTemp);
+  newtemp -= kTecoMinTemp;  // 16=0b000
+
+  remote_state &= ~kTecoTempMask;  // reinit temp
+  remote_state |= (newtemp << 8);
+}
+
+uint8_t IRTecoAC::getTemp() {
+  return ((remote_state & kTecoTempMask) >> 8) + kTecoMinTemp;
+}
+
+// Set the speed of the fan
+void IRTecoAC::setFan(uint8_t speed) {
+  uint8_t newspeed = speed;
+  switch (speed) {
+    case kTecoFanAuto:
+    case kTecoFan3:
+    case kTecoFan2:
+    case kTecoFan1:
+      break;
+    default:
+      newspeed = kTecoFanAuto;
+  }
+  remote_state &= ~kTecoFanMask;  // reinit fan
+  remote_state |= (newspeed << 4);
+}
+
+uint8_t IRTecoAC::getFan() {
+  return (remote_state & kTecoFanMask) >> 4;
+}
+
+void IRTecoAC::setMode(uint8_t mode) {
+  uint8_t newmode = mode;
+  switch (mode) {
+    case kTecoAuto:
+    case kTecoCool:
+    case kTecoDry:
+    case kTecoFan:
```

Comment:
Later on you indicate that `kTecoFan` mode doesn't have a temperature setting. I'm assuming the bits containing the temp are set to something, possibly a special value. If so, we should set that here too.

Reply:
In Fan mode, the temperature bits do not change. They keep whichever state they were when in cool or heat mode